Improve explanation on DC2Type comments

greg0ire opened the pull request:
That documentation looks quite strange right now because the only
described use-case is not very convincing.

greg0ire committed:
Make test cases easier to identify

greg0ire commented on tests/Schema/ComparatorTest.php:
```diff
@@ -1228,14 +1228,6 @@ public static function getCompareColumnComments(): iterable
-
-        self::assertFalse($this->comparator->columnsEqual($column1, $column2));
-    }
-
```
Sometimes these columns should be considered equal, and sometimes not, it depends on the platform. Locally, this test now fails for SQLServer.

morozov replied:
It looks like this test uses the platform-unaware comparator. Why does changing the platform-aware comparison logic fail this test on any platform? I believe the platform-unaware comparison behavior should be preserved as much as possible unless it's a bug.

morozov replied:
> Locally, this test now fails for SQLServer.

It's not a functional test. How can it fail for SQL Server? I see, it's the `SQLServer\ComparatorTest` that fails.

morozov replied:
> Sometimes these columns should be considered equal, and sometimes not, it depends on the platform.

Why does it depend on the platform? None of the platforms support these types natively, so I'd expect all platforms to behave the same. Although I agree that from the schema comparison perspective, these types should not be considered different (i.e. the test is invalid).

But there's still something odd: both `MySQL\ComparatorTest` and `SQLServer\ComparatorTest` cover the platform-aware scenarios for its corresponding platform but the test passes on MySQL because the generated column SQL contains different comments for different types but it fails on SQL Server because those columns don't contain any comments.

morozov replied:
For the above test to pass on SQL Server, the platform-aware logic should be implemented as:
```diff
diff --git a/src/Platforms/AbstractPlatform.php b/src/Platforms/AbstractPlatform.php
index 536b4a718..e51b40e5f 100644
--- a/src/Platforms/AbstractPlatform.php
+++ b/src/Platforms/AbstractPlatform.php
@@ -4017,9 +4017,12 @@ abstract class AbstractPlatform
      */
     public function columnsEqual(Column $column1, Column $column2): bool
     {
+        $column1Array = $this->columnToArray($column1);
+        $column2Array = $this->columnToArray($column2);
+
         if (
-            $this->getColumnDeclarationSQL('', $this->columnToArray($column1))
-            !== $this->getColumnDeclarationSQL('', $this->columnToArray($column2))
+            $this->getColumnDeclarationSQL('', $column1Array)
+            !== $this->getColumnDeclarationSQL('', $column2Array)
         ) {
             return false;
         }
@@ -4029,6 +4032,6 @@ abstract class AbstractPlatform
             return true;
         }

-        return $column1->getComment() === $column2->getComment();
+        return $column1Array['comment'] === $column2Array['comment'];
     }
 }
```
It should compare the comments that would include the column type, not only the user-defined comments.

While we're keeping the type comments in place, we should keep this test in place and apply the above patch. Even if both types produce the same SQL, they produce different comments which may be relied upon for platform-unaware comparison. We can remove this patch together with removing commented types.

greg0ire replied:
Thanks for looking into this. I applied the suggested patch before proceeding with the loosening of the comparison so that each commit stays green.

morozov replied:
> I applied the suggested patch before proceeding with the loosening of the comparison so that each commit stays green.

I just realized that the end result is the same: instead of comparing comments and types separately, the comparator now compares them in a combination. So there's no loosening of the column comparison logic in fact.

And furthermore, the comparison logic cannot be loosened because otherwise, it may result in a state where the type comments in the deployed database do not correspond to the types defined in the application-level schema.

greg0ire replied:
Oh, right. Let's close this then.

morozov commented:
Looks good but I believe it should target `3.4.x`. Shouldn't it?

greg0ire commented:
Oh right, I forgot 3.3.0 was released!

greg0ire committed:
Explain former main purpose of the comments

That documentation looks quite strange right now because the only
described use-case is not very convincing.

greg0ire committed:
Warn against relying on DC2Type comments.

greg0ire commented:
@morozov I'm reusing this branch to keep parts that pertain to documentation. I think it does not hurt clarifying further what purpose DC2Type comments serve, and how one should not rely on them.
